### ISSUE TITLE
[CoSim] Incorrect residuals in `ConvergenceCriteria` 

### ISSUE THREAD

**Issue by juancamarotti** [KratosMultiphysics/Kratos#12394]
## Issue

During coupled simulations, the initial residual of subdomains is not updated between coupling iterations. The convergence criterion is using incorrect residuals. 

## Reproducer

To reproduce, run the FSI mok example setting the structural domain echo level to 4 and looking at how the initial residual norm changes between coupling iterations (Spoiler alert: it does not change).

## Bug

The initial residual of `ConvergenceCriteria` is updated in [`ConvergenceCriteria::InitializeSolutionStep`](https://github.com/KratosMultiphysics/Kratos/blob/8956b6484ccc3160ddc79e9c943790595659f20f/kratos/solving_strategies/convergencecriterias/residual_criteria.h#L250), which is in turn called when the structural solver's `InitializeSolutionStep` is called.

In the `GaussSeidelStrongCoupledSolver`, the `ConvergenceCriteria` of the structural domain is initialized [once per time step](https://github.com/KratosMultiphysics/Kratos/blob/8956b6484ccc3160ddc79e9c943790595659f20f/applications/CoSimulationApplication/python_scripts/base_classes/co_simulation_coupled_solver.py#L182) (and [not once per coupling iteration](https://github.com/KratosMultiphysics/Kratos/blob/8956b6484ccc3160ddc79e9c943790595659f20f/applications/CoSimulationApplication/python_scripts/coupled_solvers/gauss_seidel_strong.py#L73-L122)).

FYI @philbucher @KratosMultiphysics/technical-committee

**Comment by philbucher** [comment]
nice catch!

No wonder it never really worked for us, we basically always used the other criteria

**Comment by juancamarotti** [comment]
Yes! This issue is causing slow convergence with the non-linear shell elements we are using in our coupled simulations.

**Comment by matekelemen** [comment]
The question is how to fix the bug. As far as I can tell, calling the solver's `InitializeSolutionStep` and `FinalizeSolutionStep` during coupling iterations is a no-go because it would flush the solution buffer and thus break time stepping.

A hackish solution I can think of is storing the initial residual in `ProcessInfo` or `ModelPart`'s `DataValueContainer` and manually overriding it during coupling iterations, but I don't particularly like this fix.

Any other ideas?

**Comment by juancamarotti** [comment]
After further analysis of the problem, I have concluded that this is not a CoSim issue. The bug is related to the strategy used. Let's consider the `ResidualBasedNewtonRaphsonStrategy` as an example:

1) In [ResidualBasedNewtonRaphsonStrategy::SolveSolutionStep](https://github.com/KratosMultiphysics/Kratos/blob/d828a58d28f2098645e850e15317e46a1971f82b/kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h#L920), the method [ConvergenceCriteria::InilitializeNonLinearIteration](https://github.com/KratosMultiphysics/Kratos/blob/d828a58d28f2098645e850e15317e46a1971f82b/kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h#L998) is called with the LHS and RHS even before invoking the `BuilderAndSolver`. This implies that the method is called with zero LHS and RHS vectors.

2) The second issue is that inside [ResidualCriteria](https://github.com/KratosMultiphysics/Kratos/blob/d828a58d28f2098645e850e15317e46a1971f82b/kratos/solving_strategies/convergencecriterias/residual_criteria.h#L58), the `ConvergenceCriteria::InitializeNonLinearIteration` method is not overriden. The residual is calculated just once per time step inside [ResidualCriteria::InitializeSolutionStep](https://github.com/KratosMultiphysics/Kratos/blob/d828a58d28f2098645e850e15317e46a1971f82b/kratos/solving_strategies/convergencecriterias/residual_criteria.h#L234) and not once per coupling iteration.

@KratosMultiphysics/technical-committee Why the [ConvergenceCriteria::InilitializeNonLinearIteration](https://github.com/KratosMultiphysics/Kratos/blob/d828a58d28f2098645e850e15317e46a1971f82b/kratos/solving_strategies/strategies/residualbased_newton_raphson_strategy.h#L998) is called before the `BuilderAndSolver`?

**Comment by yagofx** [comment]
Hi,

Has this been solved? I've been working on FSI simulations using shell elements and encountered problems with convergence.

**Comment by loumalouomega** [comment]
> Hi,
> 
> Has this been solved? I've been working on FSI simulations using shell elements and encountered problems with convergence.

@philbucher @matekelemen

**Comment by yagofx** [comment]
What I have seen is that at the end of the non-linear iterations in the FSI solver the residual norm is computed, so it is updated every coupling iteration. I'm no expert just started in Kratos recently for my Bachelor's 😃

**Comment by rubenzorrilla** [comment]
As far as I understand, @yagofx is not using the `CoSimulationApplication` so I don't think that his troubles have something to do with the issue @juancamarotti is referring to.
